c++/src/kj/compat/http.c++

```diff
@@ -1729,33 +1826,49 @@ static_assert(!fastCaseCmp<'f','O','o','B'>("FooB1"), "");
 static_assert(!fastCaseCmp<'f','O','o','B','1','a'>("FooB1"), "");
 
 kj::Own<kj::AsyncInputStream> HttpInputStreamImpl::getEntityBody(
-    RequestOrResponse type, HttpMethod method, uint statusCode,
+    RequestOrResponse type,
+    kj::OneOf<HttpMethod, HttpConnectMethod> method,
+    uint statusCode,
     const kj::HttpHeaders& headers) {
   KJ_REQUIRE(headerBuffer.size() > 0, "Cannot get entity body after header buffer release.");
 
+  auto isHeadRequest = method.tryGet<HttpMethod>().map([](auto& m) {
+    return m == HttpMethod::HEAD;
+  }).orDefault(false);
+
+  auto isConnectRequest = method.is<HttpConnectMethod>();
+
   // Rules to determine how HTTP entity-body is delimited:
   //   https://tools.ietf.org/html/rfc7230#section-3.3.3
-
   // #1
   if (type == RESPONSE) {
-    if (method == HttpMethod::HEAD) {
+    if (isHeadRequest) {
       // Body elided.
       kj::Maybe<uint64_t> length;
       KJ_IF_MAYBE(cl, headers.get(HttpHeaderId::CONTENT_LENGTH)) {
         length = strtoull(cl->cStr(), nullptr, 10);
       } else if (headers.get(HttpHeaderId::TRANSFER_ENCODING) == nullptr) {
-        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD request.
-        //   Propagate this fact with a 0 expected body length.
+        // HACK: Neither Content-Length nor Transfer-Encoding header in response to HEAD
+        //   request. Propagate this fact with a 0 expected body length.
         length = uint64_t(0);
       }
       return kj::heap<HttpNullEntityReader>(*this, length);
+    } else if (isConnectRequest && statusCode >= 200 && statusCode < 300) {
+      // #2 For any 2xx response to a CONNECT request... a client MUST ignore any
+      // Content-Length or Transfer-Encoding header fields received.
+      return kj::heap<HttpConnectionCloseEntityReader>(*this);
     } else if (statusCode == 204 || statusCode == 304) {
       // No body.
       return kj::heap<HttpNullEntityReader>(*this, uint64_t(0));
     }
   }
 
-  // #2 deals with the CONNECT method which is handled separately.
+  // For CONNECT requests messages, we will ignore both the Transfer-Encoding
+  // and the Content-Length here. The HttpServer will reject such requests
+  // so ignoring it here should be safe.
```

Comment:
I would not special-case CONNECT here, but just let the rest of the logic play out. If the caller is already preventing content-length and transfer-encoding headers from getting here, we end up in the same place. If the caller later changes to remove that prohibition, then this code will produce the correct kind of stream (but it's then up to the caller to use it correctly).